language/move-vm/runtime/src/debug.rs

```diff
@@ -0,0 +1,194 @@
+    }
+}
+
+impl FromStr for DebugCommand {
+    type Err = String;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        use DebugCommand::*;
+        let s = s.trim();
+        if s.starts_with(PrintStack.debug_string()) {
+            return Ok(PrintStack);
+        }
+        if s.starts_with(Step.debug_string()) {
+            return Ok(Step);
+        }
+        if s.starts_with(Continue.debug_string()) {
+            return Ok(Continue);
+        }
+        if let Some(breakpoint) = s.strip_prefix(Breakpoint("".to_owned()).debug_string()) {
+            return Ok(Breakpoint(breakpoint.to_owned()));
+        }
+        if let Some(breakpoint) = s.strip_prefix(DeleteBreakpoint("".to_owned()).debug_string()) {
+            return Ok(DeleteBreakpoint(breakpoint.to_owned()));
+        }
+        if s.starts_with(PrintBreakpoints.debug_string()) {
+            return Ok(PrintBreakpoints);
+        }
+        Err(format!(
+            "Unrecognized command: {}\nAvailable commands: {}",
+            s,
+            Self::commands()
+                .iter()
+                .map(|command| command.debug_string())
+                .collect::<Vec<_>>()
+                .join(", ")
+        ))
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct DebugContext {
+    breakpoints: BTreeSet<String>,
+    should_take_input: bool,
+}
+
+impl DebugContext {
+    pub(crate) fn new() -> Self {
+        Self {
+            breakpoints: BTreeSet::new(),
+            should_take_input: true,
+        }
+    }
+
+    pub(crate) fn debug_loop(
+        &mut self,
+        function_desc: &Function,
+        locals: &Locals,
+        pc: u16,
+        instr: &Bytecode,
+        resolver: &Loader,
+        interp: &Interpreter,
+    ) {
+        let instr_string = format!("{:?}", instr);
+        let function_string = function_desc.pretty_string();
+        let breakpoint_hit = self.breakpoints.contains(&function_string)
+            || self
+                .breakpoints
+                .iter()
+                .any(|bp| instr_string[..].starts_with(bp.as_str()));
+
+        if self.should_take_input || breakpoint_hit {
+            self.should_take_input = true;
+            if breakpoint_hit {
+                let bp_match = self
+                    .breakpoints
+                    .iter()
+                    .find(|bp| instr_string.starts_with(bp.as_str()))
+                    .unwrap()
+                    .clone();
+                println!(
+                    "Breakpoint {} hit with instruction {}",
+                    bp_match, instr_string
+                );
+            }
+            println!(
+                "function >> {}\ninstruction >> {:?}\nprogram counter >> {}",
+                function_string, instr, pc
+            );
+            loop {
+                print!("> ");
+                std::io::stdout().flush().unwrap();
+                let mut input = String::new();
+                match io::stdin().read_line(&mut input) {
```

Comment:
Interesting approach! So we're effectively instrumenting each instruction with a call into the debugging loop, which will then block on stdin for user input. I think this would be a great and very usable at least for now. Wondering if we would prefer offering a set of debugging/stepping interfaces from the VM instead in the future.

@dariorussi I'd love to hear your opinions on this as a short-term solution.

Reply:
this seems fine to me and it's a way to have some minimal debugging ability so it's all good.... :)